### PR TITLE
Updated Render Subpasses sample

### PR DESCRIPTION
**Pull request by ZandroFargnoli**
## Description

Updated the `Render Subpasses` sample:

* Renamed to `Subpasses` 
* Tutorial references and screenshots updated 
* Gbuffer overflow now shows a delta on S10s and S20s

## General Checklist:

Please ensure the following points are checked:

- [x] My code follows the [coding style](https://github.com/KhronosGroup/Vulkan-Samples/tree/master/CONTRIBUTING.md#Code-Style)
- [x] I have reviewed file [licenses](https://github.com/KhronosGroup/Vulkan-Samples/tree/master/CONTRIBUTING.md#Copyright-Notice-and-License-Template)
- [x] I have commented any added functions (in line with Doxygen)
- [x] I have commented any code that could be hard to understand
- [x] My changes do not add any new compiler warnings
- [x] My changes do not add any new validation layer errors or warnings
- [x] I have used existing framework/helper functions where possible
- [x] My changes build and run on Windows, Linux, macOS and Android. Otherwise I have [documented any exceptions](https://github.com/KhronosGroup/Vulkan-Samples/tree/master/CONTRIBUTING.md#General-Requirements)

## Sample Checklist

If your PR contains a new or modified sample, these further checks must be carried out *in addition* to the General Checklist:
- [x] I have tested the sample on at least one compliant Vulkan implementation
- [x] If the sample is vendor-specific, I have [tagged it appropriately](https://github.com/KhronosGroup/Vulkan-Samples/tree/master/CONTRIBUTING.md#General-Requirements)
- [x] I have stated on what implementation the sample has been tested so that others can test on different implementations and platforms
- [x] Any dependent assets have been merged and published in downstream modules

### CODE REVIEW

File: samples/performance/subpasses/subpasses_tutorial.md [hunk 1]
```diff
@@ -54,59 +54,58 @@ vec4 world_w = inv_view_proj * clip;
...
+Some framebuffer attachments in the sample (such as _depth_, _albedo_, and _normal_), are cleared at the beginning of the render pass, written by the geometry subpass, read by the lighting subpass, and discarded at the end of the render pass. If the GPU has enough memory available to store them on tile memory, there is no need to write them back to external memory. Actually, there is not even a need to allocate them at all.
 
 ![Non-transient attachments](images/transient-attachments.jpg)
 
+In practice, their [image usage](https://www.khronos.org/registry/vulkan/specs/1.1-extensions/man/html/VkImageUsageFlagBits.html) needs to be specified as `TRANSIENT` and their [memory](https://www.khronos.org/registry/vulkan/specs/1.1-extensions/man/html/VkMemoryPropertyFlagBits.html) needs to be `LAZILY_ALLOCATED`. Failing to set these flags properly will lead to an increase of [fragment jobs](https://community.arm.com/developer/tools-software/graphics/b/blog/posts/mali-bifrost-family-performance-counters) as the GPU will need to write them back to external memory. As you can see in the above screenshot, we see roughly a double in fragment jobs per second (from `56/s` to `113/s`).
```
Comment (gary-sweet) [inline samples/performance/subpasses/subpasses_tutorial.md:102]: Is this correct? I would expect transient buffers to affect bandwidth, but not the number of fragment jobs.

File: samples/performance/subpasses/subpasses_tutorial.md [hunk 2]
```diff
@@ -54,59 +54,58 @@ vec4 world_w = inv_view_proj * clip;
...
-
-From the sample perspective, the best way to verify whether two subpasses are merged or not is to compare the physical tiles (`PTILES`) counter by switching between the sub-passes and the render passes technique. Two fused sub-passes will need half the number of `PTILES` needed by two render passes, indeed comparing the following screenshot with the previous one, roughly half the number of tiles are used (`0.5` vs `1.1`) and about `70%` of bandwidth is saved.
+Furthermore, in order to be merged, subpasses are required to use at most 128-bit per pixel of *tile buffer color storage*, although some of the more recent GPUs such as Mali-G72 increase this to up to 256-bits per pixel. G-buffers requiring more color storage than this can be used at the expense of needing smaller tiles during fragment shading, which can reduce overall throughput and increase bandwidth reading tile lists.
```
Comment (r-potter) [inline samples/performance/subpasses/subpasses_tutorial.md]: I think this sentence could use a tweak to clarify that the limits are Mali specific. Something like:

> On Mali GPUs, subpass merging is limited to subpasses which utilize at most 128-bits per pixel of *tile buffer color storage*, although some of the more recent Mali GPUs such as Mali-G72 relax this limit to 256-bits per pixel.

File: samples/performance/subpasses/subpasses_tutorial.md [hunk 3]
```diff
@@ -117,7 +116,7 @@ In practice, their [image usage](https://www.khronos.org/registry/vulkan/specs/1
 
 **Do**
 
-* Use sub-passes.
+* Use subpasses.
 * Use a 128-bit G-buffer budget for color.
```
Comment (r-potter) [inline samples/performance/subpasses/subpasses_tutorial.md]: This limit appears to be Mali-specific. Maybe reword to a more general "aim to keep your tile buffer budget small"

**Review by r-potter — DISMISSED** [review]
The readme could use a clarifying pass to make obvious where the stated limits are Mali limits.

Also discussed in the call. I'd like to see a frame time counter added so that the tradeoffs between bandwidth reduction and FPS are clear. Currently the use of subpasses appears to be a huge positive, when in practice they typically involve balancing power usage, bandwidth and FPS. Some text in the readme highlighting that tradeoff would also be helpful. I don't mind if we do that separately (or I can do it as a follow up MR if preferred).

Otherwise looks good to me. Tested on various Samsung Mali and Adreno devices.

**Comment by AlejandroCosin** [comment]
I tested the sample Subpasses in an Asus Chromebook R13 with a PowerVR GX6250 GPU, works correctly.

**Review by AlejandroCosin — DISMISSED** [review]

File: samples/performance/subpasses/subpasses_tutorial.md (hunk 1 above)
Comment (ZandroFargnoli) [reply samples/performance/subpasses/subpasses_tutorial.md:102]: AFAIK when it comes to transient vs non-transient, as long as the image is storeOp=DONT_CARE, there's rarely any change to the BW. Transient only directly impacts the memory footprint which has nothing to do with bandwidth consumption.

Comment (gary-sweet) [reply samples/performance/subpasses/subpasses_tutorial.md:102]: Yes, I should have written memory, not bandwidth. My bad.
I still can't see how it affects the number of fragment jobs though.

Comment (ZandroFargnoli) [reply samples/performance/subpasses/subpasses_tutorial.md:102]: I think the confusion might be in the definition of fragment jobs

Comment (ZandroFargnoli) [reply samples/performance/subpasses/subpasses_tutorial.md:102]: In Mali we sometimes schedule fragment jobs to do some setup/metadata-management, but these are often tiny single-thread type jobs poking some value into memory.

Comment (ZandroFargnoli) [reply samples/performance/subpasses/subpasses_tutorial.md:102]: What is the Broadcom equivalent of this? Are you seeing the fragment jobs not increasing at all with the respective setup?

Comment (gary-sweet) [reply samples/performance/subpasses/subpasses_tutorial.md:102]: For us, fragment jobs represent the number of rendering tasks that the GPU has been given (roughly equivalent to one per  subpass in this example). Neither transient buffers or changes to the storeOp can affect this for us.

I could understand if changing from renderpasses to subpasses changed this, but the fact that it's transient buffers confuses me.

This sentence in particular: "Failing to set these flags properly will lead to an increase of [fragment jobs] ... as the GPU will need to write them back to external memory".

Comment (ZandroFargnoli) [reply samples/performance/subpasses/subpasses_tutorial.md:102]: Do you see no change in the fragment job counter?

Comment (ZandroFargnoli) [reply samples/performance/subpasses/subpasses_tutorial.md:102]: Can I suggest, seeing as this is just to rename the sample, that we move this to an issue to be resolved at a later stage. This paragraph exists in master.

Comment (gary-sweet) [reply samples/performance/subpasses/subpasses_tutorial.md:102]: No change in the counter for us, no.

> move this to an issue to be resolved at a later stage

Sure. No problem.

**Review by AlejandroCosin — CHANGES_REQUESTED** [review]
In desktop, for some reason, the Radio Button widgets are not clickable: I've to offset the mouse to see the button highlight and press it (all radio buttons work fine if you manage to click them, it's just the mouse offset). Tried in two different computers.

**Comment by PaoloJovon-arm** [comment]
> In desktop, for some reason, the Radio Button widgets are not clickable: I've to offset the mouse to see the button highlight and press it (all radio buttons work fine if you manage to click them, it's just the mouse offset). Tried in two different computers.

Hello @AlejandroCosin! Could you please try applying #156 to see if it fixes this issue?

**Comment by AlejandroCosin** [comment]
Hi @PaoloJovon-arm

Tried your changes in GlfwWindow::get_content_scale_factor() and the issue is 100% fixed!

**Comment by PaoloJovon-arm** [comment]
@AlejandroCosin Glad to hear. Thanks for testing it!

**Review by gary-sweet — DISMISSED** [review]
This change on its own looks fine.
The text still needs some work (as discussed) and I would also like to see a frame-time counter added, but I'm ok with those being done in a future change if necessary.

**Review by AlejandroCosin — DISMISSED** [review]

**Review by r-potter — APPROVED** [review]

**Review by gary-sweet — APPROVED** [review]

**Review by aurolou — APPROVED** [review]
